### ISSUE TITLE
Phase 41.2 — RobustnessVerifier: Certified Robustness Bounds & Formal Verification

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#826]
# Phase 41.2 — RobustnessVerifier

## Overview

The **RobustnessVerifier** provides formal and probabilistic certification of model robustness — moving beyond empirical attacks to **provable guarantees**. It implements randomized smoothing (Cohen et al., 2019), convex relaxation bounds (Wong & Kolter, 2018), and abstract interpretation methods to certify that a model's prediction is stable within a defined perturbation region.

## Architecture

```
┌─────────────────────────────────────────────────────────────┐
│                     RobustnessVerifier                       │
│                                                              │
│  ┌───────────────────┐  ┌────────────────────────────────┐  │
│  │ Certification      │  │ Bound Computation Engine       │  │
│  │ Strategy Selector  │  │                                │  │
│  │                    │  │  ┌────────────────────────┐    │  │
│  │  Randomized        │  │  │ Interval Bound         │    │  │
│  │  Smoothing         │──│  │ Propagation (IBP)      │    │  │
│  │                    │  │  └────────────────────────┘    │  │
│  │  Convex            │  │  ┌────────────────────────┐    │  │
│  │  Relaxation        │──│  │ CROWN / α-CROWN        │    │  │
│  │                    │  │  └────────────────────────┘    │  │
│  │  Abstract          │  │  ┌────────────────────────┐    │  │
│  │  Interpretation    │──│  │ DeepPoly / ZonoDomain  │    │  │
│  │                    │  │  └────────────────────────┘    │  │
│  │  Lipschitz         │  │  ┌────────────────────────┐    │  │
│  │  Analysis          │──│  │ Spectral Norm Bounds   │    │  │
│  │                    │  │  └────────────────────────┘    │  │
│  └───────────┬────────┘  └──────────────┬─────────────────┘  │
│              │                           │                    │
│  ┌───────────▼───────────────────────────▼─────────────────┐ │
│  │           Certificate Store                              │ │
│  │  Per-sample certificates • Certified radius • Stats      │ │
│  └──────────────────────────────────────────────────────────┘ │
└─────────────────────────────────────────────────────────────┘
```

## TypeScript Interface

```typescript
import { Tensor, Model } from '@asi-build/core';

interface CertificationConfig {
  method: 'randomized_smoothing' | 'ibp' | 'crown' | 'deeppoly' | 'lipschitz';
  norm: 'L2' | 'Linf';
  epsilon: number;
  confidence?: number;          // for randomized smoothing (default 0.999)
  numSamples?: number;          // Monte Carlo samples for smoothing (default 100000)
  alpha?: number;               // significance level (default 0.001)
  batchSize?: number;
  timeout?: number;             // per-sample timeout in ms
}

interface Certificate {
  input: Tensor;
  predictedClass: number;
  certified: boolean;           // is the prediction provably robust?
  certifiedRadius: number;      // maximum provably safe perturbation radius
  method: string;
  confidence: number;           // statistical confidence (for randomized smoothing)
  computeTimeMs: number;
  lowerBound?: Tensor;          // certified lower bounds on logits
  upperBound?: Tensor;          // certified upper bounds on logits
}

interface CertificationReport {
  modelId: string;
  method: string;
  norm: string;
  epsilon: number;
  totalSamples: number;
  certifiedCount: number;
  certifiedAccuracy: number;    // fraction provably robust at epsilon
  avgCertifiedRadius: number;
  medianCertifiedRadius: number;
  radiusHistogram: Map<number, number>;
  perClassCertified: Map<number, number>;
  computeTimeMs: number;
}

interface RobustnessVerifier {
  // Randomized smoothing (Cohen et al., 2019)
  certifySmoothing(model: Model, input: Tensor, config: CertificationConfig): Promise<Certificate>;
  
  // Interval Bound Propagation
  certifyIBP(model: Model, input: Tensor, epsilon: number): Promise<Certificate>;
  
  // CROWN linear relaxation
  certifyCROWN(model: Model, input: Tensor, epsilon: number): Promise<Certificate>;
  
  // Lipschitz-based certification
  certifyLipschitz(model: Model, input: Tensor, epsilon: number): Promise<Certificate>;
  
  // Batch certification
  batchCertify(model: Model, dataset: Tensor[], labels: number[], config: CertificationConfig): Promise<CertificationReport>;
  
  // Certified radius computation
  computeCertifiedRadius(model: Model, input: Tensor, method: string): Promise<number>;
  
  // Verify specific claim
  verifyClaim(model: Model, input: Tensor, label: number, epsilon: number, norm: string): Promise<boolean>;
}
```

## Randomized Smoothing Implementation

```python
class RandomizedSmoothing:
    """Cohen et al. (2019) - Certified Adversarial Robustness via Randomized Smoothing."""
    
    def __init__(self, base_model, sigma: float, n_samples: int = 100000, alpha: float = 0.001):
        self.base_model = base_model
        self.sigma = sigma
        self.n_samples = n_samples
        self.alpha = alpha
    
    def predict(self, x: Tensor) -> int:
        """Return the smoothed classifier's prediction."""
        counts = self._sample_under_noise(x, self.n_samples)
        return counts.argmax().item()
    
    def certify(self, x: Tensor) -> Tuple[int, float]:
        """Certify robustness and return (predicted_class, certified_radius)."""
        # Step 1: Identify top class with high confidence
        counts = self._sample_under_noise(x, self.n_samples)
        top_class = counts.argmax().item()
        top_count = counts[top_class].item()
        
        # Step 2: Clopper-Pearson lower bound on p_A
        p_lower = proportion_confint(top_count, self.n_samples, 
                                      alpha=2*self.alpha, method='beta')[0]
        
        if p_lower <= 0.5:
            return top_class, 0.0  # Cannot certify
        
        # Step 3: Certified radius = σ * Φ^(-1)(p_A)
        radius = self.sigma * norm.ppf(p_lower)
        return top_class, radius
    
    def _sample_under_noise(self, x: Tensor, n: int) -> Tensor:
        """Count predictions under Gaussian noise."""
        counts = torch.zeros(self.num_classes)
        for batch in chunks(range(n), self.batch_size):
            noisy = x.repeat(len(batch), 1, 1, 1) + torch.randn_like(...) * self.sigma
            preds = self.base_model(noisy).argmax(dim=1)
            for p in preds:
                counts[p] += 1
        return counts
```

## Configuration

```yaml
robustness_verifier:
  default_method: "randomized_smoothing"
  
  randomized_smoothing:
    sigma: 0.25                  # noise standard deviation
    n_samples_certify: 100000
    n_samples_predict: 1000
    alpha: 0.001                 # significance level
    batch_size: 1000
  
  ibp:
    propagation_mode: "interval"
    merge_layers: true
  
  crown:
    alpha_crown: true            # use optimizable linear bounds
    beta_crown: true             # branch and bound
    iteration_limit: 100
    lr_alpha: 0.1
  
  lipschitz:
    method: "spectral_norm"
    power_iterations: 10
    
  certification:
    timeout_per_sample_ms: 30000
    parallel_workers: 4
    save_certificates: true
```

## Testing Strategy

```typescript
describe('RobustnessVerifier', () => {
  describe('Randomized Smoothing', () => {
    it('should certify correctly classified clean samples', async () => {
      const cert = await verifier.certifySmoothing(model, cleanInput, { 
        method: 'randomized_smoothing', sigma: 0.25, numSamples: 10000 
      });
      expect(cert.certified).toBe(true);
      expect(cert.certifiedRadius).toBeGreaterThan(0);
    });
    
    it('should return radius=0 for uncertain predictions', async () => {
      const cert = await verifier.certifySmoothing(model, ambiguousInput, config);
      expect(cert.certifiedRadius).toBe(0);
    });
    
    it('should satisfy soundness: no attack within radius succeeds', async () => {
      const cert = await verifier.certifySmoothing(model, input, config);
      if (cert.certified && cert.certifiedRadius > 0) {
        const attackResult = await attacker.pgd(model, input, label, {
          epsilon: cert.certifiedRadius * 0.99, numSteps: 100
        });
        expect(attackResult.success).toBe(false);
      }
    });
  });
  
  describe('IBP', () => {
    it('should produce valid interval bounds', async () => {
      const cert = await verifier.certifyIBP(model, input, 0.01);
      expect(cert.lowerBound).toBeDefined();
      expect(cert.upperBound).toBeDefined();
    });
  });
  
  describe('Batch Certification', () => {
    it('should report certified accuracy', async () => {
      const report = await verifier.batchCertify(model, testSet, labels, config);
      expect(report.certifiedAccuracy).toBeGreaterThanOrEqual(0);
      expect(report.certifiedAccuracy).toBeLessThanOrEqual(1);
    });
  });
});
```

## Acceptance Criteria

- [ ] Randomized smoothing produces valid certified radii (statistically sound)
- [ ] IBP bounds are provably correct (lower bound ≤ true activation ≤ upper bound)
- [ ] CROWN achieves tighter bounds than IBP on standard benchmarks
- [ ] Certified accuracy at ε=0.5 (L2) on CIFAR-10 matches published baselines
- [ ] Lipschitz estimation converges with increasing power iterations
- [ ] Batch certification scales to 10K+ samples
- [ ] Certificates are stored with full provenance and reproducibility metadata
- [ ] Soundness property holds: no attack within certified radius succeeds

## References

- Cohen et al. (2019) — Certified Adversarial Robustness via Randomized Smoothing
- Wong & Kolter (2018) — Provable Defenses via Convex Outer Adversarial Polytope
- Salman et al. (2019) — Provably Robust Deep Learning via Adversarially Trained Smoothed Classifiers
- Gowal et al. (2019) — Scalable Verified Training for Provably Robust Image Classification
- Xu et al. (2020) — Automatic Perturbation Analysis (CROWN / α-CROWN)

---
*Sub-phase of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824)*
